New block: Downloads Chart

ryelle opened the pull request:
Fixes #32 — Adds a new block for the download chart. This uses google charts like the current theme, with some slight styling updates (fonts, colors).

Note: This is using `jQuery.getJSON` + `jsonp` to get the data, since this endpoint doesn't have `Access-Control-Allow-Origin: *` and I was getting cors issues when using wp.apiFetch. If that's easy to fix, this can be updated to not rely on jQuery.

### Screenshots

This can show up to 260 days (columns) of data, though the chart looks best when it's much smaller than that… Perhaps a future iteration could use a line chart instead?

Brand new theme, only up for 2 days:

![Screen Shot 2024-05-09 at 18 51 02](https://github.com/WordPress/wporg-theme-directory/assets/541093/86888707-942c-441a-a4ba-a09a46e1b350)

Slightly older theme, 17 days:

![Screen Shot 2024-05-09 at 18 50 14](https://github.com/WordPress/wporg-theme-directory/assets/541093/f4e7c41d-8891-4775-9443-97c0a9b40aec)

Full 260 days of data (Twenty Twenty-Three)

![Screen Shot 2024-05-09 at 18 50 43](https://github.com/WordPress/wporg-theme-directory/assets/541093/e11ab87c-6d1d-4d58-a6f0-f89b5608b4ce)

Full 260 days of data, less popular theme, with hover state

![Screenshot 2024-05-09 at 6 59 21 PM](https://github.com/WordPress/wporg-theme-directory/assets/541093/5b55cbf7-94af-4994-ba9a-54fb565ea07b)

### How to test the changes in this Pull Request:

1. View a theme (if working locally, make sure it has a real theme slug)
2. There should be a graph on the page
3. It should match the graph on the current site, in terms of data

ryelle committed:
New block: Downloads Chart

StevenDufresne commented on source/wp-content/themes/wporg-themes-2024/patterns/single.php:
```diff
@@ -96,12 +96,14 @@
-		<!-- wp:paragraph -->
-		<p>[graph]</p>
+		<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"wporg-themes/meta","args":{"key":"active-installs"}}}}} -->
+		<p>Active Installations</p>
```
This likely needs localization as well?

ryelle replied:
No, this text doesn't matter — it's there for devex really, so we know what _will_ be there. It could be empty. The actual text is populated by the block binding `"content":{"source":"wporg-themes/meta","args":{"key":"active-installs"}}`.

I've used the [same pattern in the meta section here](https://github.com/WordPress/wporg-theme-directory/blob/a89dca1c0fb1bf72086902e24fa7be27302de5af/source/wp-content/themes/wporg-themes-2024/patterns/single.php#L51-L77).

StevenDufresne commented on source/wp-content/themes/wporg-themes-2024/src/theme-downloads/index.php:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	$theme_post = get_post( $block->context['postId'] );
+	$theme = wporg_themes_theme_information( $theme_post->post_name );
```
You may want to check this return value more specifically in case it returns an [error](https://github.com/WordPress/wordpress.org/blob/ed2884a3cfbfb45bf0ba327c67643d96ca5bfeeb/wordpress.org/public_html/wp-content/plugins/theme-directory/class-themes-api.php#L502).

ryelle replied:
Done in https://github.com/WordPress/wporg-theme-directory/pull/54/commits/e005f29c1c1362d2f9ef40901144519d86d9d689, also catching any errors from the API in https://github.com/WordPress/wporg-theme-directory/pull/54/commits/5997c9f3cf814ff8e410a8baf5fe4bcd5eda722b.

StevenDufresne commented on source/wp-content/themes/wporg-themes-2024/src/theme-downloads/index.php:
```diff
@@ -0,0 +1,63 @@
+	$theme = wporg_themes_theme_information( $theme_post->post_name );
+
+	$attributes = array(
+		'theme' => $theme->slug,
```
Should this be escaped?

ryelle replied:
Done for all the attributes in https://github.com/WordPress/wporg-theme-directory/pull/54/commits/15e3713c99b92e7013455b61e55841e262d0d7dc

StevenDufresne commented on source/wp-content/themes/wporg-themes-2024/src/theme-downloads/view.js:
```diff
@@ -0,0 +1,82 @@
+		} );
+
+		new google.visualization.ColumnChart( element ).draw( data, {
+			title: 'Density of Precious Metals, in g/cm^3',
```
Do we need this title? :)

ryelle replied:
Nope! Whoops. I was trying out different settings and this comes from the chart docs 🤦🏻

ryelle replied:
Removed in https://github.com/WordPress/wporg-theme-directory/pull/54/commits/8a107d47eaa8bbd697a47ddd13a1ab285dce43f9

StevenDufresne commented on source/wp-content/themes/wporg-themes-2024/src/theme-downloads/view.js:
```diff
@@ -0,0 +1,82 @@
+			},
+			hAxis: {
+				textStyle: {
+					color: '#1e1e1e',
```
We could grab this from the body if we wanted to:
```
getComputedStyle(document.body).getPropertyValue('--wp--preset--color--charcoal-1');
```

ryelle replied:
Good idea! I've set up a colors object for all the colors I use: https://github.com/WordPress/wporg-theme-directory/pull/54/commits/a0a89821c344b95ee749335b5a66e8594c87602c

StevenDufresne commented:
I tested the PR; working as expected. I added a few minor comments.

As you note, we can probably make some nice improvements to these charts for readability but I also agree it's not relevant to this PR or issue.

👍

ryelle committed:
Escape attributes when building string

ryelle committed:
Check the theme object for errors

ryelle committed:
Attach jQuery dependency to the google chart

This isn't exactly correct, it's a dependency of the view script, but including it in block.json was triggering a "doing it wrong" message about adding `defer` data. This avoids that warning.

ryelle committed:
Remove testing title config

ryelle committed:
Get the colors from custom properties

ryelle committed:
Switch to `ajax()` with explicit `jsonp` type to prevent jQuery Migrate warning, catch any errors

StevenDufresne reviewed (approved):
LGTM.

adamwoodnz commented on source/wp-content/themes/wporg-themes-2024/src/theme-downloads/view.js:
```diff
@@ -0,0 +1,97 @@
+		data.addColumn( 'string', labelDate );
+		data.addColumn( 'number', labelValue );
+
+		Object.entries( downloads ).forEach( ( [ key, value ] ) => {
```
It's unlikely, but if there was no error and the ajax return value is anything other than an object, `Object.entries( downloads )` will result in a TypeError:

![Screenshot 2024-05-13 at 10 56 21 AM](https://github.com/WordPress/wporg-theme-directory/assets/1017872/96b29fd1-1298-478d-9818-dcafd52e1f85)

Could just initialize `downloads` as an empty object.

ryelle committed:
Initialize downloads as empty object
